### PR TITLE
Configure a Bootable Jar cluster application to use DNS_PING protocol

### PR DESCRIPTION
**Pull request by yersan**
A blog post about how to configure DNS ping protocol for a Bootable Jar cluster application

### CODE REVIEW

**Review by jmesnil — COMMENTED** [review]
That's a great introduction to bootable jar script to specialise an application

File: _posts/2021-03-15-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+date:   2021-03-15
+tags: wildfly galleon bootable jar bootable-jar clustering jkube dns-ping
+author: yborgess
+---
+
+= Configuring DNS ping for a WildFly Bootable JAR cluster application
+
+== Introduction
+
+We recently described in this https://www.wildfly.org/news/2021/02/01/Bootable-jar-jkube-clustering-openshift[post] details about how to deploy and configure a WildFly Bootable JAR cluster application on OpenShift. By default, a Bootable JAR cluster application uses the http://www.jgroups.org/manual5/index.html#_kube_ping[KUBE_PING] protocol for clustering discovery mechanism. This protocol is based on the ability to ask the Kubernetes API for information about what are the pods that match certain criteria and then create a JGroups cluster based on this information. This protocol could not be a viable or an ideal solution for your scenario, for example, if you cannot grant a specific permission on your Kubernetes cluster to use KUBE_PING, or if your requirement is to make your application agnostic to the Kubernetes API.
```
[the remark this answers is not in the document]
Comment (yersan) [reply _posts/2021-03-15-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc]: Much better, thanks for the suggestion. Done.

File: _posts/2021-03-15-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+In this guide we will explain how you can configure the same Bootable JAR cluster application but this time by using http://www.jgroups.org/manual5/#_dns_ping[DNS_PING] protocol.
```
[the remark this answers is not in the document]
Comment (yersan) [reply _posts/2021-03-15-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc]: Fixed.

File: _posts/2021-03-15-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+[NOTE]
```
[the remark this answers is not in the document]
Comment (yersan) [reply _posts/2021-03-15-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc]: Yes, I agree. I moved it to the beginning and added a note about the follow-up on the previous article.

File: _posts/2021-03-15-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        </build>
+    </profile>
+</profiles>
+----
+
+== Build, deploy and verify the demo application
+
+Now let us build and deploy our Bootable JAR application. We assume you have a cluster running and you have already logged into it.
+
+1. If you have already cloned the demo application, checkout the *dns_ping* branch.
```
[the remark this answers is not in the document]
Comment (yersan) [reply _posts/2021-03-15-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc]: I added how to clone it and avoid assume it is already clone.

**Review by yersan — COMMENTED** [review]
Thank you @jmesnil for the review, I followed up on all your suggestions

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 5]
```diff
@@ -0,0 +1,162 @@
+---
+layout: post
+date:   2021-03-10
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:3]: Can you change this date and the file name to 2021-03-12 please?

The WF 23 release announcement will go out tomorrow and I want it to be the first post on the list that day. But I don't want this getting buried under it either; it's really good and I want it on top. Merging it Friday will achieve that.

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+tags: wildfly galleon bootable jar bootable-jar clustering jkube dns-ping
+author: yborgess
+---
+
+= Configuring DNS ping on a WildFly Bootable JAR cluster application
+
+[NOTE]
+====
+This post is a continuation of https://www.wildfly.org/news/2021/02/01/Bootable-jar-jkube-clustering-openshift[WildFly Bootable JAR cluster application with JKube openshift-maven-plugin] blog post, it assumes you are familiar with the https://github.com/yersan/wildfly-clustering-demo.git[demo application] we used for that.
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:12]: s/continuation of/continuation of my/g

s/post, it/post. It/g

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+====
+
+== Introduction
+
+We recently described in this https://www.wildfly.org/news/2021/02/01/Bootable-jar-jkube-clustering-openshift[post] details about how to deploy and configure a WildFly Bootable JAR cluster application on OpenShift. By default, a Bootable JAR cluster application uses the http://www.jgroups.org/manual5/index.html#_kube_ping[KUBE_PING] protocol for clustering discovery mechanism. This protocol is based on the ability to query the Kubernetes API to determine which pods match certain criteria and then create a JGroups cluster based on this information. This protocol could not be a viable or an ideal solution for your scenario, for example, if you cannot grant a specific permission on your Kubernetes cluster to use KUBE_PING, or if your requirement is to make your application agnostic to the Kubernetes API.
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:17]: s/for clustering discovery/for its clustering discovery/g

s/could not/may not/g

s/scenario, for example/scenario; for example/g

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+In this guide we will explain how you can configure the same Bootable JAR cluster application but this time by using the http://www.jgroups.org/manual5/#_dns_ping[DNS_PING] protocol.
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:19]: s/time by using/time using/g

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+* *Configure the JGroups subsystem to use DNS_PING.* We need to change the default JGroups subsystem configuration of our Bootable JAR application. We will execute a CLI script at build time to configure the JGroups protocol stack, removing the default KUBE_PING protocol and adding the DNS_PING. One important detail about the DNS_PING configuration is the value we will use for the *dns_query* property. This value specifies the DNS name of the Kubernetes ping service the protocol is going to use to discover other cluster members.
+
+To sum up, we will need a Kubernetes service with a matching label selector to define which pods will be part of our cluster, and we will tell to the JGroups subsystem that we want to use such a service to locate other possible cluster members.
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:29]: s/tell to the/tell the/g

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+== Define a ping service
+
+The ping service is just an additional Kubernetes service resource. We are using https://www.eclipse.org/jkube/docs/openshift-maven-plugin[JKube openshift-maven-plugin] to deploy our application, this maven plugin allows us to specify any additional resource we want to deploy. What we need to do is to create our additional resources under *src/main/jkube/* and JKube will create the resource for us when deploying the application.
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:33]: s/are using/are using the/g

s/application, this/application. This/g

s/is to create/is create/g

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    description: The JGroups ping service for clustering discovery.
+spec:
+  clusterIP: None
+  publishNotReadyAddresses: true
+  selector:
+    app: wildfly-clustering-demo
+----
+
+The *selector* specifies the name of our application. Without any additional modification, the name of our application is the maven group id of our project.
+We have also configured the service with *publishNotReadyAddresses: true*. This configuration specifies that when there is a DNS lookup by using this service, the result will contain all pods matching the service’s label selector, even those which are not yet in a ready state. We need it to properly discover the pods IP when they are not yet in the ready state.
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:51]: s/need it to/need this in order to/g

Perhaps add "We want the JGroups cluster to form before the pods are receiving end user requests."

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+We have also added *clusterIP: None* in the service spec, it means this service will not assign a cluster IP through which clients can connect to the all pods backing it, instead it will return an IP to connect directly to the pod.
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:52]: s/spec, it/spec. This/g

s/to the all/to all the/g

s/it, instead/it; instead/g

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+As a second step, we need to configure the JGroups stack to use the DNS_PING protocol. We have to instruct the  https://github.com/wildfly-extras/wildfly-jar-maven-plugin[wildfly-jar-maven-plugin] to execute the https://github.com/yersan/wildfly-clustering-demo/blob/dns-ping/src/main/scripts/configure_dns_ping.cli[following script]:
+
+----
+/subsystem=jgroups/stack=tcp/protocol=kubernetes.KUBE_PING:remove
+/subsystem=jgroups/stack=tcp/protocol=dns.DNS_PING:add(add-index=0)
+/subsystem=jgroups/stack=tcp/protocol=dns.DNS_PING/property=dns_query:add(value=${env.DNS_PING_SERVICE_NAME})
+/subsystem=jgroups/stack=tcp/protocol=dns.DNS_PING/property=async_discovery_use_separate_thread_per_request:add(value=true)
+----
+
+The script removes the existing *kubernetes.KUBE_PING* protocol and adds the *dns.DNS_PING*. The value of the *dns_query* property will be retrieved from an environment variable named *DNS_PING_SERVICE_NAME*. The value of this environment variable is specified under the *<env>* section on the JKube maven plugin configuration in our pom.xml.
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:65]: s/section on the/section of the/g

File: _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+The maven _openshift_ profile looks like as https://github.com/yersan/wildfly-clustering-demo/blob/dns-ping/pom.xml#L53-L100[follows]:
```
Comment (bstansberry) [inline _posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc:67]: s/like as/as

**Review by bstansberry — CHANGES_REQUESTED** [review]
Very nice content!

**Review by yersan — COMMENTED** [review]
Thank you @jmesnil and @bstansberry for the additional review and feedback, I followed your suggestions and added a minimal conclusion paragraph at the end.

**Pull request merged**